plugins/cn0540.c

```diff
@@ -0,0 +1,574 @@
+				"vshift_buf"));
+	vsensor_log = GTK_TEXT_VIEW(gtk_builder_get_object(builder,
+				"vsensor_log"));
+	vsensor_buf = GTK_TEXT_BUFFER(gtk_builder_get_object(builder,
+				"vsensor_buf"));
+
+	make_widget_update_signal_based(iio_widgets, num_widgets);
+
+	g_signal_connect(G_OBJECT(&tgbtn_shutdown->toggle_button), "toggled",
+		G_CALLBACK(monitor_shutdown), NULL);
+	g_signal_connect(G_OBJECT(&tgbtn_fda->toggle_button), "toggled",
+		G_CALLBACK(monitor_fda), NULL);
+	g_signal_connect(G_OBJECT(&tgbtn_fda_mode->toggle_button), "toggled",
+		G_CALLBACK(monitor_fda_mode), NULL);
+	g_signal_connect(G_OBJECT(btn_get_sw_ff), "clicked",
+		G_CALLBACK(monitor_sw_ff), &btn_get_sw_ff);
+	g_signal_connect(G_OBJECT(calib_btn),"clicked",
+		G_CALLBACK(calib),&calib_btn);
+	g_signal_connect(G_OBJECT(read_btn),"clicked",
+		G_CALLBACK(read_vshift),&read_btn);
+	g_signal_connect(G_OBJECT(write_btn),"clicked",
+		G_CALLBACK(write_vshift),&write_btn);
+	g_signal_connect(G_OBJECT(readvsensor_btn),"clicked",
+		G_CALLBACK(read_vsensor),&readvsensor_btn);
+
+	iio_update_widgets(iio_widgets, num_widgets);
+}
+
+static void cn0540_init()
+{
+	int idx;
+
+	gtk_toggle_button_toggled(&tgbtn_shutdown->toggle_button);
+	gtk_toggle_button_toggled(&tgbtn_fda->toggle_button);
+	gtk_toggle_button_toggled(&tgbtn_fda_mode->toggle_button);
+	gtk_button_clicked(btn_get_sw_ff);
+	gtk_button_clicked(calib_btn);
+
+	if (iio_voltage_mon != NULL)
+		g_timeout_add_seconds(1, (GSourceFunc)update_voltages,
+				     iio_voltage_mon);
+	else {
+		for(idx = 0; idx < NUM_ANALOG_PINS; idx++) {
+			gtk_text_buffer_set_text(voltage_buffer[idx],
+						"INACTIVE", -1);
+		}
+	}
+}
+
+static GtkWidget *cn0540_plugin_init(struct osc_plugin *plugin,
+				     GtkWidget *notebook, const char *ini_fn)
+{
+	GtkBuilder *builder;
+
+	builder = gtk_builder_new();
+
+	ctx = osc_create_context();
+	if (!ctx)
+		return NULL;
+
+	if (osc_load_glade_file(builder, "cn0540") < 0)
+		return NULL;
+
+	block_diagram_init(builder, 1, "CN0540.jpg");
+
+	cn0540_get_channels();
+	cn0540_plugin_interface_init(builder);
+	cn0540_init();
+
+	return cn0540_panel;
+}
+
+static bool cn0540_identify(const struct osc_plugin *plugin)
+{
+	/* Use the OSC's IIO context just to detect the devices */
+	struct iio_context *osc_ctx = get_context_from_osc();
+
+	/* Get the iio devices */
+	iio_adc = iio_context_find_device(osc_ctx, ADC_DEVICE);
+	iio_dac = iio_context_find_device(osc_ctx, DAC_DEVICE);
+	iio_gpio = iio_context_find_device(osc_ctx, GPIO_CTRL);
+	iio_voltage_mon = iio_context_find_device(osc_ctx, VOLTAGE_MONITOR_1);
+	if (!iio_voltage_mon)
+		iio_voltage_mon = iio_context_find_device(osc_ctx,
+							  VOLTAGE_MONITOR_2);
+
+	if (!iio_adc || !iio_dac || !iio_gpio) {
+		printf("Could not find expected iio devices\n");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void context_destroy(struct osc_plugin *plugin, const char *ini_fn)
+{
+	g_source_remove_by_user_data(ctx);
```

Comment:
Is `g_source_remove_by_user_data()` really needed? Or should it be `g_source_remove_by_user_data(iio_voltage_mon)`?